File: solutions/python/flower-field/1/flower_field.py
# Dedicated to my Shree DR.MDD

ADJACENT = [complex(a, b) for a in (-1, 0, 1) for b in (-1, 0, 1) if a or b]

def annotate(garden: list[str]) -> list[str]:
    h = len(garden)
    w = len(garden[0]) if garden else 0

    if not all(len(r) == w for r in garden):
        raise ValueError("The board is invalid with current input.")

    blossom = set()
    for j, stripe in enumerate(garden):
        for i, ch in enumerate(stripe):
            if ch == "*":
                blossom.add(complex(i, j))
            elif ch != " ":
                raise ValueError("The board is invalid with current input.")

    def bloom(p: complex) -> str:
        if p in blossom:
            return "*"
        cnt = sum((p + d) in blossom for d in ADJACENT)
        return str(cnt) if cnt else " "

    return [
        "".join(bloom(complex(i, j)) for i in range(w))
        for j in range(h)
    ]

File: solutions/julia/flower-field/1/flower-field.jl
# Dedicated to Shree DR.MDD

function annotate(field)
    rcount = length(field)
    ccount = rcount == 0 ? 0 : length(field[1])
    output = copy(field)

    for r in 1:rcount
        chs = collect(output[r])
        for c in 1:ccount
            if field[r][c] == '*'
                continue
            end

            mines = 0
            for dr in -1:1, dc in -1:1
                rr, cc = r + dr, c + dc
                if dr == 0 && dc == 0
                    continue
                end
                if 1 ≤ rr ≤ rcount && 1 ≤ cc ≤ ccount
                    field[rr][cc] == '*' && (mines += 1)
                end
            end

            if mines > 0
                chs[c] = Char('0' + mines)
            end
        end
        output[r] = join(chs)
    end

    output
end
